### ISSUE TITLE
detecting non-renderable URLs

### ISSUE THREAD

**Issue by maxogden** [atom/electron#1253]
If I load a non-renderable URL, such as a `.zip` or something that would prompt a 'download as' prompt in a browser, atom-shell renders a blank page.

The problem I'm running in to is I am not sure how to detect that something is non-renderable.

Here's a gif of what I mean:

![atom-shell-load-events](https://cloud.githubusercontent.com/assets/39759/6680872/7f1e522a-cc1a-11e4-9ed5-09fdd44d4ae0.gif)

Here's the source code of the test program: https://gist.github.com/maxogden/891f9f69be667e11fa5e

I've noticed that the `did-stop-loading` event fires before `did-finish-load`, and in the `did-stop-loading` callback I get `false` from `WebContents.isLoading()`

My problem is that I have no way of knowing if `did-stop-loading` (correction: I mean `did-finish-load`) will ever fire, as it never fires for non-renderable content. But there doesn't seem to be a way to know if the content is non-renderable through the API.

**Comment by bwin** [comment]
> My problem is that I have no way of knowing if did-stop-loading will ever fire, as it never fires for non-renderable content.

I think you mean "if `did-finish-load` will ever fire". If you don't, it's already too late for me (I was already in bed, but couldn't sleep anyway).

Also +1 on your prompt.

TL;DR: keep looking for a "real" solution.

So, I can (in a really, really hacky way) determine if the loading failed and all works well, if I test like you showed in your gif (by manually closing the window). If I try to automate that (close the window when it's non-renderable) I have to use a delay before closing the window (for some reason).
So even if you would consider doing it this ugly way (not recommended), the delay would be a no-go for your use-case.
Also, the more I run it, the more I think the delay should be even longer to not give any false-positives.

So I wouldn't consider this a solution to your problem (at all), but nonetheless I wanted to share what I tried/did/whatever. Maybe it's best to not look at it?! https://gist.github.com/bwin/973ce7fb8396124a5469
Also, I didn't try any other URLs then the 2 you provided. You shouldn't do it my way anyway, so it doesn't matter...

Also not nice but still better (if there currently is no right way to do it) would be to send a HEAD request and look at the `Content-type`, maybe.

@zcbenz can probably tell you how to do it properly.

**Comment by bwin** [comment]
Yeah, this should definitely fire an event. Either `did-fail-load` or something new.
(Some people may even want to get a "download as"-dialog...)

**Comment by zcbenz** [comment]
There is a [will-navigate](https://github.com/atom/atom-shell/blob/master/docs/api/browser-window.md#event-will-navigate) event can be used to prevent loading urls.

**Comment by maxogden** [comment]
@zcbenz `will-navigate` never fires in my example above, I think because I am creating a window, loading 1 URL, and then closing the window.

My problem is that I have no way of knowing when atom-shell rendered a blank page using the currently available events. What do you think of the `did-fail-load` event that @bwin proposed?

**Comment by maxogden** [comment]
Maybe chromium has some sort of event for e.g. 'render type'? For example, if I load a `.js` file I get this:

![screen shot 2015-03-18 at 1 31 12 pm](https://cloud.githubusercontent.com/assets/39759/6718735/149d767e-cd73-11e4-93da-3e101cbb0da0.png)

So somewhere in chromium it knows how to display .js files (e.g. it generates an html page with a <span> etc.). Could this be surfaced through an event in atom-shell?

Also it would be *really* cool to have access to the HTTP data (especially the response data). Is this possible? Figuring out the 'render type' of a URI could probably be done pretty reliably with access to the HTTP response data because I'm guessing Chromium uses content-type as the major factor for deciding how to render something

![screen shot 2015-03-18 at 1 32 46 pm](https://cloud.githubusercontent.com/assets/39759/6718765/59afef30-cd73-11e4-9182-4faf483a893d.png)

**Comment by zcbenz** [comment]
I agree we should have a way to handle this.

It is hard to hook before Chromium renders a file though, Chromium just hard coded those things somewhere deep in the source code.

**Comment by maxogden** [comment]
@zcbenz @deepak1556 awesome!!!
